helix-term/src/ui/statusline.rs

```diff
@@ -0,0 +1,336 @@
+                + viewport
+                    .width
+                    .saturating_sub(context.buffer.right.width() as u16),
+            viewport.y,
+            &context.buffer.right,
+            context.buffer.right.width() as u16,
+        );
+
+        // Center of the status line.
+
+        let element_ids = &editor.config().status_line.center;
+        element_ids
+            .iter()
+            .map(|element_id| Self::get_render_function(*element_id))
+            .for_each(|render| render(context, write_center));
+
+        // Width of the empty space between the left and center area and between the center and right area.
+        let spacing = 1u16;
+
+        let edge_width = context
+            .buffer
+            .left
+            .width()
+            .max(context.buffer.right.width()) as u16;
+        let center_max_width = viewport.width - (2 * edge_width + 2 * spacing);
+        let center_width = center_max_width.min(context.buffer.center.width() as u16);
+
+        surface.set_spans(
+            viewport.x + viewport.width / 2 - center_width / 2,
+            viewport.y,
+            &context.buffer.center,
+            center_width,
+        );
+    }
+
+    fn append(buffer: &mut Spans, text: String, base_style: &Style, style: Option<Style>) {
+        buffer.0.push(Span::styled(
+            text,
+            style.map_or(*base_style, |s| (*base_style).patch(s)),
+        ));
+    }
+
+    fn get_render_function<F>(element_id: StatusLineElementID) -> impl Fn(&mut RenderContext, F)
+    where
+        F: Fn(&mut RenderContext, String, Option<Style>) + Copy,
+    {
+        match element_id {
+            helix_view::editor::StatusLineElement::Mode => Self::render_mode,
+            helix_view::editor::StatusLineElement::Spinner => Self::render_lsp_spinner,
+            helix_view::editor::StatusLineElement::FileName => Self::render_file_name,
+            helix_view::editor::StatusLineElement::FileEncoding => Self::render_file_encoding,
+            helix_view::editor::StatusLineElement::FileType => Self::render_file_type,
+            helix_view::editor::StatusLineElement::Diagnostics => Self::render_diagnostics,
+            helix_view::editor::StatusLineElement::Selections => Self::render_selections,
+            helix_view::editor::StatusLineElement::Position => Self::render_position,
+        }
+    }
+
+    fn render_mode<F>(context: &mut RenderContext, write: F)
+    where
+        F: Fn(&mut RenderContext, String, Option<Style>) + Copy,
+    {
+        let visible = context.focused;
+
+        write(
+            context,
+            format!(
+                " {} ",
+                match context.doc.mode() {
+                    Mode::Insert if visible => "INS",
+                    Mode::Select if visible => "SEL",
+                    Mode::Normal if visible => "NOR",
+                    // If not focused, explicitly leave an empty space instead of returning None.
+                    _ => "   ",
```

Comment:
```suggestion
                    _ if !visible => "   ",
```
I think we want this enum to be exhaustive so if we add a variant next time we won't miss this.

Reply:
Hmm, very good suggestion, but with the change, it seems like the compiler doesn't recognize the `match` arms as exhaustive anymore:

```rust
match context.doc.mode() {
    Mode::Insert if visible => "INS",
    Mode::Select if visible => "SEL",
    Mode::Normal if visible => "NOR",
    // If not focused, explicitly leave an empty space instead of returning None.
    _ if !visible => "   ",
}
```

```
error[E0004]: non-exhaustive patterns: `Normal`, `Select` and `Insert` not covered
   --> helix-term/src/ui/statusline.rs:166:23
    |
166 |                 match context.doc.mode() {
    |                       ^^^^^^^^^^^^^^^^^^ patterns `Normal`, `Select` and `Insert` not covered
    |
   ::: /home/user/Projects/helix/helix-view/src/document.rs:35:5
    |
35  |     Normal = 0,
    |     ------ not covered
36  |     Select = 1,
    |     ------ not covered
37  |     Insert = 2,
    |     ------ not covered
    |
    = help: ensure that all possible cases are being handled, possibly by adding wildcards or more match arms
    = note: the matched value is of type `helix_view::document::Mode`
```

Reply:
Sorry to just jump in, but couldn't you do something like
```
format!(
    " {} ",
    if visible {
        match {
            Mode::Insert => "INS",
            ...,
        } else {
            "    "
        }
    });
```
?
This way you have the variant checks, as well as pulling the visibility test outside the match.

Reply:
I initially didn't want to clutter the code too much, but it doesn't look that bad after all, and it's a useful change. I committed this in 0f53b17d525e75bdc7c71e90cdb7578fc8fb38d4.